### PR TITLE
Merging LoRA weights to form base model weight

### PR DESCRIPTION
**Pull request by timothylimyl**
Allowing for all the benefits of inference using base model weights such as quantisation. I believe there could be many more improvement made to the base model in the future.

I have tested that output using newly merged weights is good on `llm.int8`

Reference to  [Issue #261 ](https://github.com/Lightning-AI/lit-llama/issues/261)

### CODE REVIEW

File: scripts/convert_lora_weights.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    dt = getattr(torch, dtype, None)
+    if not isinstance(dt, torch.dtype):
+        raise ValueError(f"{dtype} is not a valid dtype.")
+    dtype = dt
+
+    print("Loading model ...", file=sys.stderr)
+    t0 = time.time()
+
+    with (lazy_load(checkpoint_path) as pretrained_checkpoint,
+          lazy_load(lora_path) as adapter_checkpoint):
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:72]:
```suggestion
          lazy_load(lora_path) as lora_checkpoint):
```

File: scripts/convert_lora_weights.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        with EmptyInitOnDevice(
+                device=fabric.device, dtype=dtype
+        ), lora(r=rank, alpha=16, dropout=0.05, enabled=True):
+            model = LLaMA.from_name(name)
+
+            # 1. Load the pretrained weights
+            model.load_state_dict(pretrained_checkpoint, strict=False)
+            # 2. Load the fine-tuned adapter weights
+            model.load_state_dict(adapter_checkpoint, strict=False)
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:84]:
```suggestion
            # 2. Load the fine-tuned lora weights
            model.load_state_dict(lora_checkpoint, strict=False)
```

File: scripts/convert_lora_weights.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        raise ValueError(f"{dtype} is not a valid dtype.")
+    dtype = dt
+
+    print("Loading model ...", file=sys.stderr)
+    t0 = time.time()
+
+    with (lazy_load(checkpoint_path) as pretrained_checkpoint,
+          lazy_load(lora_path) as adapter_checkpoint):
+        name = llama_model_lookup(pretrained_checkpoint)
+        rank = lora_model_lookup(adapter_checkpoint)
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:74]:
```suggestion
        rank = lora_model_lookup(lora_checkpoint)
```

File: scripts/convert_lora_weights.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    dtype: str = "bfloat16",
+) -> None:
+    """Merges lora weights to base model.
+
+    Args:
+        accelerator: The hardware to run on. Possible choices are:
+            ``"cpu"``, ``"cuda"``, ``"mps"``, ``"gpu"``, ``"tpu"``, ``"auto"``.
+        lora_path: Path to the checkpoint with trained LoRA weights, which are the output of
+            `finetune_lora.py`.
+        checkpoint_path: The checkpoint path to load.
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:50]: can you add dtype too?

File: scripts/convert_lora_weights.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+def del_lora_state_dict(model: nn.Module):
+    key_to_delete = []
+    base_model_dict = model.state_dict()
+    for k in base_model_dict:
+        if "lora_" in k:
+            key_to_delete.append(k)
+    for del_key in key_to_delete:
+        del base_model_dict[del_key]
+    del model
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:26]: deleting the model inside this function is strange. The `del_lora_state_dict` should just delete the lora state dict, freeing up memory can be done outside

File: scripts/convert_lora_weights.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+from lit_llama import LLaMA
+from lit_llama.utils import EmptyInitOnDevice, lazy_load, llama_model_lookup
+from lit_llama.lora import lora
+
+def del_lora_state_dict(model: nn.Module):
+    key_to_delete = []
+    base_model_dict = model.state_dict()
+    for k in base_model_dict:
+        if "lora_" in k:
+            key_to_delete.append(k)
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:23]: This could be a simple list comprehension

File: scripts/convert_lora_weights.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            # 1. Load the pretrained weights
+            model.load_state_dict(pretrained_checkpoint, strict=False)
+            # 2. Load the fine-tuned adapter weights
+            model.load_state_dict(adapter_checkpoint, strict=False)
+
+    print(f"Time to load model: {time.time() - t0:.02f} seconds.", file=sys.stderr)
+
+    model.eval()
+    base_model_dict = del_lora_state_dict(model)
+    log_name = str(lora_path).split("/")[-1][:-4] + "-lora-merged-weights.pth"
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:90]: Here you could keep the Path type and do `lora_path.name.with_stem(...)`

File: scripts/convert_lora_weights.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    save_path = Path(log_path / log_name)
+    print("Saving LoRA to base model weights ...")
+    torch.save(base_model_dict, save_path)
+    print(f"Model saved at {str(save_path)}")
+
+
+if __name__ == "__main__":
+    from jsonargparse import CLI
+
+    torch.set_float32_matmul_precision("high")
```
Comment (carmocca) [inline scripts/convert_lora_weights.py:101]: This is not necessary for this script
```suggestion
```

File: scripts/convert_lora_weights.py (hunk 7 above)
Comment (timothylimyl) [reply scripts/convert_lora_weights.py:90]: cool, did not know that feature existed!

**Comment by timothylimyl** [comment]
@carmocca sorry about some of the rookie mistakes, I have cleaned it up. Thanks for the suggestions.

**Review by lantiga — APPROVED** [review]
Looks great, I'm good to merge

Thank you @timothylimyl!

**Pull request merged**